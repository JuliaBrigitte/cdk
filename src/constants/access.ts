/**
 * An enum representing different types of application access.
 */
export enum AccessScope {
  /**
   * For when you want your application to be accessible to the world (0.0.0.0/0).
   */
  PUBLIC = "Public",

  /**
   * For when you want to restrict your application's access to a list of CIDR ranges.
   */
  RESTRICTED = "Restricted",

  /**
   * For when you want to restrict your application's access to the VPC.
   */
  INTERNAL = "Internal",
}
