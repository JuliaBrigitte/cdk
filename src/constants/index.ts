export * from "./regex-pattern";
export * from "./stage";
