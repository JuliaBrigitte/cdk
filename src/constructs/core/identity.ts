export interface Identity {
  stack: string;
  stage: string;
  app: string;
}
