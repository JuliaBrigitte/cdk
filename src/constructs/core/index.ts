export * from "./identity";
export * from "./mappings";
export * from "./migrating";
export * from "./parameters";
export * from "./ssm";
export * from "./stack";
