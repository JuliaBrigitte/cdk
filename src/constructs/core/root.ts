import type { StageSynthesisOptions } from "aws-cdk-lib";
import { App } from "aws-cdk-lib";
import type { CloudAssembly } from "aws-cdk-lib/cx-api";
import { RiffRaffYamlFile } from "../../utils/riff-raff-yaml-file";

// Deliberately not called `GuApp` as `app` is overloaded (as is `stack`, and `stage` ðŸ˜…)
export class GuRoot extends App {
  override synth(options?: StageSynthesisOptions): CloudAssembly {
    new RiffRaffYamlFile(this).synth();
    return super.synth(options);
  }
}
