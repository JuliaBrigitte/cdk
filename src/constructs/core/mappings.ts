import { CfnMapping } from "@aws-cdk/core";
import type { GuStack } from "./stack";

export interface GuStageDependentValue {
  variableName: string;
  codeValue: unknown;
  prodValue: unknown;
}

export class GuStageMapping extends CfnMapping {
  constructor(scope: GuStack, id: string = "stage-mapping") {
    super(scope, id);
  }
}
