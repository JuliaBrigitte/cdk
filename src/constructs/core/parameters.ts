import type { CfnParameterProps } from "@aws-cdk/core";
import { CfnParameter } from "@aws-cdk/core";
import { Stage, Stages } from "../../constants";
import type { GuStack } from "./stack";

export type GuParameterProps = CfnParameterProps;

export type GuNoTypeParameterProps = Omit<GuParameterProps, "type">;

export class GuParameter extends CfnParameter {
  constructor(scope: GuStack, id: string, props: GuParameterProps) {
    super(scope, id, props);
  }
}

export class GuStringParameter extends GuParameter {
  constructor(scope: GuStack, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, { ...props, type: "String" });
  }
}

export class GuStageParameter extends GuParameter {
  constructor(scope: GuStack) {
    super(scope, "Stage", {
      type: "String",
      description: "Stage name",
      allowedValues: Stages,
      default: Stage.CODE,
    });
  }
}

export class GuStackParameter extends GuParameter {
  constructor(scope: GuStack) {
    super(scope, "Stack", {
      type: "String",
      description: "Name of this stack",
      default: "deploy",
    });
  }
}

// TODO: Is there a way of removing default props if they weren't implemented before?
//       Should that be allowed even if it is possible?
export class GuInstanceTypeParameter extends GuParameter {
  constructor(scope: GuStack, id: string = "InstanceType", props: GuParameterProps = {}) {
    super(scope, id, {
      type: "String",
      description: "EC2 Instance Type",
      default: "t3.small",
      ...props,
    });
  }
}

export class GuSSMParameter extends GuParameter {
  constructor(scope: GuStack, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, {
      noEcho: true,
      ...props,
      type: "AWS::SSM::Parameter::Value<String>",
    });
  }
}

export class GuSubnetListParameter extends GuParameter {
  constructor(scope: GuStack, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, { ...props, type: "List<AWS::EC2::Subnet::Id>" });
  }
}

export class GuVpcParameter extends GuParameter {
  constructor(scope: GuStack, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, {
      ...props,
      type: "AWS::EC2::VPC::Id",
    });
  }
}

export class GuAmiParameter extends GuParameter {
  constructor(scope: GuStack, id: string, props: GuNoTypeParameterProps) {
    super(scope, id, {
      ...props,
      type: "AWS::EC2::Image::Id",
    });
  }
}
