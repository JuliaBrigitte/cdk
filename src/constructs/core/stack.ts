import type { App, StackProps } from "@aws-cdk/core";
import { Stack, Tags } from "@aws-cdk/core";
import { GuStackParameter, GuStageParameter } from "./parameters";

export interface GuStackProps extends StackProps {
  // This limits GuStack to supporting a single app.
  // In the future, support for stacks with multiple apps may be required
  app: string;
  migratedFromCloudFormation?: boolean;
}

export class GuStack extends Stack {
  private readonly _stage: GuStageParameter;
  private readonly _stack: GuStackParameter;
  private readonly _app: string;

  get stage(): string {
    return this._stage.valueAsString;
  }

  get stack(): string {
    return this._stack.valueAsString;
  }

  get app(): string {
    return this._app;
  }

  migratedFromCloudFormation: boolean;

  protected addTag(key: string, value: string, applyToLaunchedInstances: boolean = true): void {
    Tags.of(this).add(key, value, { applyToLaunchedInstances });
  }

  // eslint-disable-next-line custom-rules/valid-constructors -- GuStack is the exception as it must take an App
  constructor(app: App, id: string, props: GuStackProps) {
    super(app, id, props);

    this.migratedFromCloudFormation = !!props.migratedFromCloudFormation;

    this._stage = new GuStageParameter(this);
    this._stack = new GuStackParameter(this);
    this._app = props.app;

    this.addTag("Stack", this.stack);
    this.addTag("Stage", this.stage);
    this.addTag("App", props.app);
  }
}
