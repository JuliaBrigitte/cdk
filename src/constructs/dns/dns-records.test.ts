import "@aws-cdk/assert/jest";
import "../../utils/test/jest";
import { SynthUtils } from "@aws-cdk/assert";
import { Duration } from "@aws-cdk/core";
import { Stage } from "../../constants";
import { simpleGuStackForTesting } from "../../utils/test";
import { GuCname, GuDnsRecordSet, RecordType } from "./dns-records";

describe("The GuDnsRecordSet construct", () => {
  it("should create the correct resources with minimal config", () => {
    const stack = simpleGuStackForTesting();
    new GuDnsRecordSet(stack, "TestRecord", {
      name: "banana.example.com",
      recordType: RecordType.CNAME,
      resourceRecords: ["apple.example.com"],
      ttl: Duration.hours(1),
    });
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });
  it("should use the exact logical id that is passed in", () => {
    const stack = simpleGuStackForTesting();
    new GuDnsRecordSet(stack, "ThisExactLogicalId", {
      name: "banana.example.com",
      recordType: RecordType.CNAME,
      resourceRecords: ["apple.example.com"],
      ttl: Duration.hours(1),
    });
    expect(stack).toHaveResourceOfTypeAndLogicalId("Guardian::DNS::RecordSet", "ThisExactLogicalId");
  });
});

describe("The GuCname construct", () => {
  it("should create the correct resources with minimal config", () => {
    const stack = simpleGuStackForTesting();
    new GuCname(stack, "TestRecord", {
      domainNameProps: {
        [Stage.CODE]: { domainName: "xyz.code-guardian.com" },
        [Stage.PROD]: { domainName: "xyz.prod-guardian.com" },
      },
      app: "my-test-app",
      resourceRecord: "apple.example.com",
      ttl: Duration.hours(1),
    });
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
  });
});
