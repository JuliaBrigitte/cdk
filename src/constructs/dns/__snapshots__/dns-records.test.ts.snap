// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The GuCname construct should create the correct resources with minimal config 1`] = `
Object {
  "Mappings": Object {
    "mytestapp": Object {
      "CODE": Object {
        "domainName": "xyz.code-guardian.com",
      },
      "PROD": Object {
        "domainName": "xyz.prod-guardian.com",
      },
    },
  },
  "Parameters": Object {
    "Stage": Object {
      "AllowedValues": Array [
        "CODE",
        "PROD",
      ],
      "Default": "CODE",
      "Description": "Stage name",
      "Type": "String",
    },
  },
  "Resources": Object {
    "TestRecord": Object {
      "Properties": Object {
        "Name": Object {
          "Fn::FindInMap": Array [
            "mytestapp",
            Object {
              "Ref": "Stage",
            },
            "domainName",
          ],
        },
        "RecordType": "CNAME",
        "ResourceRecords": Array [
          "apple.example.com",
        ],
        "Stage": Object {
          "Ref": "Stage",
        },
        "TTL": 3600,
      },
      "Type": "Guardian::DNS::RecordSet",
    },
  },
}
`;

exports[`The GuDnsRecordSet construct should create the correct resources with minimal config 1`] = `
Object {
  "Parameters": Object {
    "Stage": Object {
      "AllowedValues": Array [
        "CODE",
        "PROD",
      ],
      "Default": "CODE",
      "Description": "Stage name",
      "Type": "String",
    },
  },
  "Resources": Object {
    "TestRecord": Object {
      "Properties": Object {
        "Name": "banana.example.com",
        "RecordType": "CNAME",
        "ResourceRecords": Array [
          "apple.example.com",
        ],
        "Stage": Object {
          "Ref": "Stage",
        },
        "TTL": 3600,
      },
      "Type": "Guardian::DNS::RecordSet",
    },
  },
}
`;
